3.x draft (for ci check only)

halx99 opened the pull request:
## Describe your changes

- [x] Remove all deprecated stubs
- [x] Implement drawArraysInstanced and make instanced draw more common use
- [x] box2d-3.x
    - [x] Make Box2dTest of `cpp-tests` works on box2d v3
    - [x] Make Box2dTestBed of `cpp-tests` works on box2d v3
- [x] Remove chipmunk
    - [x] Remove `core/physics` based on chipmunk
    - [x] Remove ChipmunkTest from `cpp-tests`
    - [x] Remove ChipmuniTestbed from `cpp-tests`
- ~~Change underlaying color type to Color4F aka `ax::Color`~~
- [x] Improve DrawNode, don't create backend::Buffer always
- [x] Upgrade iOS deploy target: 12.0 ==> 13.0
- [x] Upgrade android minsdk: 17(Android 4.2) ==> 21(Android 5.1)
- [x] Upgrade android ndk: r23d ==> r27c
- [x] Due to android,ios toolchain update to support c++20 full features, `std::views::split` can be use to improve performance
- [x] Remove  3rdparty ghc filesystem, since both ios13 and android5.1 support `std::filesystem`
- [x] Build and run wasm64 support, current need enable chrome flag: `chrome://flags/#enable-webassembly-memory64`
- [ ] drop rapidjson
- [ ] re-select json lib: yyjson > boost.json > rapidjson
    simdjson for parse only and simdjson build in-progress: https://github.com/simdjson/experimental_json_builder
    finally solution: `yyjson`  + `simdjson`
- [x] `Color4B` ==> `Color32`
- [ ] Remove `Color3B`
- [ ] drop tolua, use sol2
- [ ] Reimplement core/physics with box2d v3, not all API compatible with axmol-2.x
## Issue ticket number and link


## Checklist before requesting a review
### For each PR
- [ ] Add Copyright if it missed:   
      - `"Copyright (c) 2019-present Axmol Engine contributors (see AUTHORS.md)."`
- [ ] I have performed a self-review of my code.
       
   Optional:
   - [ ] I have checked readme and add important infos to this PR.
   - [ ] I have added/adapted some tests too.
          
### For core/new feature PR
- [ ] I have checked readme and add important infos to this PR.
- [ ] I have added thorough tests.

## Note

*Currently, the branch `3.x` is in draft mode, that means it's not stable and will force-push frequently*

halx99 committed:
Remove deprecated stubs

halx99 committed:
Implement drawArraysInstanced and make instanced draw more common use

halx99 committed:
Bump 3.0.0

halx99 committed:
Mark DrawNode::clear as virtual function

halx99 committed:
Improve ImGuiPresenter, prevent mouse scroll event when handled

smilediver commented:
From outside it would be easier to track changes being made to `3.x` branch if they all went through PRs like in the `dev` branch. They would also get CI check automatically, and there would be an opportunity for reviews and comments.

halx99 commented:
currently, 3.x is a draft branch or you can consider it is a temp branch, I don't like make PR one by one to waste much time. I create this PR is only for non-win32 platform ci-check, because I'm compile and test on my windows computer.

halx99 committed:
Make box2d 3.0 works and remove all chipmunk related

aismann commented:
Physics should be an own test project like fairgui.
@halx99 
What are you mean?

halx99 commented:
> Physics should be an own test project like fairgui. @halx99 What are you mean?

I think physics test in cpp-tests is ok

halx99 committed:
Change underlaying color type to Color4F aka ax::Color

halx99 committed:
Improve DrawNode, don't create backend::Buffer always

halx99 committed:
Fix ci

halx99 committed:
Improve code style

aismann commented:
@halx99 
Please have a look on the Performance Candy Mix test:
- drawOrder: true it very slow (2.2 is faster) and closing get this:

Assertion failed!
Program: ..._GIT\axmol\build\bin\cpp-tests\Debug\cpp-tests.exe
File: D:\__GIT\axmol\core\renderer\CustomCommand.cpp
Line: 221
Expression: _vertexBuffer

I getting this also on other tests when I switched drawOrder: to 'true'

halx99 commented:
> @halx99 Please have a look on the Performance Candy Mix test:
> 
> * drawOrder: true it very slow (2.2 is faster) and closing get this:
> 
> Assertion failed!
> 
> Program: ..._GIT\axmol\build\bin\cpp-tests\Debug\cpp-tests.exe File: D:__GIT\axmol\core\renderer\CustomCommand.cpp Line: 221
> 
> Expression: _vertexBuffer

![image](https://github.com/user-attachments/assets/a42ccdf7-64b4-4770-92cc-2a773d669853)

Tested, works well, and no assertion

aismann commented:
Click on this text plz: it switched the drawOrder
![image](https://github.com/user-attachments/assets/92885fa5-b2c5-4f52-a567-3a6ed2fa6707)
It raised if you switched back to false

here another example:
![image](https://github.com/user-attachments/assets/22b3cf67-0186-4b24-95b3-a599b1caafdb)

halx99 committed:
Fix assetion fail in drawOrder enabled

halx99 commented:
> Click on this text plz: it switched the drawOrder ![image](https://private-user-images.githubusercontent.com/8652787/373867695-92885fa5-b2c5-4f52-a567-3a6ed2fa6707.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgxMjgxMTMsIm5iZiI6MTcyODEyNzgxMywicGF0aCI6Ii84NjUyNzg3LzM3Mzg2NzY5NS05Mjg4NWZhNS1iMmM1LTRmNTItYTU2Ny0zYTZlZDJmYTY3MDcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMDVUMTEzMDEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTg1MGI4NWEzMmMwZTk0ODViNDRmZjNkMDEzNDYzOTU3MDU5NTA5MTkwZmVkODc4YTgwODcwNjRmZTkxNTRlYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.38v5YsAt7KUf2jDzPghdAXheKivWAXn72JojFH-AGDs) It raised if you switched back to false
> 
> here another example: ![image](https://private-user-images.githubusercontent.com/8652787/373867917-22b3cf67-0186-4b24-95b3-a599b1caafdb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgxMjgxMTMsIm5iZiI6MTcyODEyNzgxMywicGF0aCI6Ii84NjUyNzg3LzM3Mzg2NzkxNy0yMmIzY2Y2Ny0wMTg2LTRiMjQtOTViMy1hNTk5YjFjYWFmZGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMDVUMTEzMDEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWE5MjlmOGRlYzA2MGQyNThjZDhiMmJkOTcxOGQ0NjRkNDc3MmQyZDI3YjIzM2ZlZTYyYWY1OGZjZTJhZTBiNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Co2q-v7VU9zF8oJdue-nGHCI9RqD5Yb5Cz4Ro8ZvAk0)

Assertion failed was fixed.

aismann commented:
Yeah. Works now.
Please have a look on the release comparing between 2.2/3.0 on Performance Candy Mix test again:
2.2 release version drawNode  true:
Seems faster (on my machine >50 FPS always)
![image](https://github.com/user-attachments/assets/f1aed79a-2efa-4afb-837f-1b938f840b99)

3.0  release version drawNode  true:
![image](https://github.com/user-attachments/assets/c4b5499f-2019-4a90-bd0c-6c0b03dcbc74)


And the drawing looks very bad like pixel jumps around whatever on Performance Candy Mix

halx99 commented:
On my machine, release build, both 2.2  & 3.0 is 60fps

aismann commented:
> On my machine, release build, both 2.2 & 3.0 is 60fps

Your machine is faster. You cant see this behavior.
What about the crazy drawing ? Looks it the same candy?

You have switch the drawOrder =>true (on the both release biulds)?

halx99 commented:
I also saw the crazy drawing

aismann commented:
> On my machine, release build, both 2.2 & 3.0 is 60fps

You have switch the drawOrder =>true (on the both release biulds)?

aismann commented:
> I also saw the crazy drawing

Maybe the _customCommandPoint stuff  has an bug?

halx99 commented:
> > On my machine, release build, both 2.2 & 3.0 is 60fps
> 
> You have switch the drawOrder =>true (on the both release biulds)?

Yes, I will check in Debug build

halx99 commented:
> > I also saw the crazy drawing
> 
> Maybe the _customCommandPoint stuff has an bug?

Not sure, there is no different with other custom commands

aismann commented:
> > > On my machine, release build, both 2.2 & 3.0 is 60fps
> > 
> > 
> > You have switch the drawOrder =>true (on the both release biulds)?
> 
> Yes, I will check in Debug build

You see it better comparing the release builds

aismann commented:
> 
> Not sure, there is no different with other custom commands

May this is the reason?
 Improve DrawNode, don't create backend::Buffer always

halx99 commented:
> > Not sure, there is no different with other custom commands
> 
> May this is the reason? Improve DrawNode, don't create backend::Buffer always

No, I revert the changes, still crazing

halx99 commented:
I have test debug build, 3.0: 33fps, 2.2: 27fps

aismann commented:
> No, I revert the changes, still crazing

Maybe a (Color) typo anywhere?

aismann commented:
> I have test debug build, 3.0: 33fps, 2.2: 27fps

Here: It was my mistake... Sorry
Debug 3.0:
![image](https://github.com/user-attachments/assets/e2ab4aa2-6340-482e-83a5-a726a9755be1)

Debug 2.2:
![image](https://github.com/user-attachments/assets/246e957f-8883-4839-906f-be4209c176b4)

halx99 commented:
debug build:


2.2:
![image](https://github.com/user-attachments/assets/1da2d8f5-1851-441a-bf9d-07e8d9f1639a)

3.0:
![image](https://github.com/user-attachments/assets/d6c6a350-f896-4be4-a0a2-8f70af1a542d)

halx99 committed:
Make DrawNodeTest CandyMixEffect behavior same with 2.x

halx99 commented:
The crazing behavior caused by timer, I revert to use box2d timer, then same with 2.2

aismann commented:
I will send a Release comparing tomorrow.

> The crazing behavior caused b timer, I revert to use box2d timer, then same with 2.2

Maybe a Helpdesk info about the "Dont use: ` utils::getTimeInMilliseconds()` is also needed?

halx99 commented:
> > No, I revert the changes, still crazing
> 
> Maybe a (Color) typo anywhere?

sure

halx99 committed:
Tidy vertex format

aismann commented:
@halx99 
How can I make a PR for 3.x?
Seems its not open (I cant send PR)

halx99 commented:
> @halx99 How can I make a PR for 3.x? Seems its not open (I cant send PR)

it's not ready for accept PR yet

aismann commented:
> > @halx99 How can I make a PR for 3.x? Seems its not open (I cant send PR)
> 
> it's not ready for accept PR yet

I will only add missing stuff.
e.g the cpp template has 
- deprecrated AXLOG
- Color4B

or you can change it of course

halx99 commented:
> > > @halx99 How can I make a PR for 3.x? Seems its not open (I cant send PR)
> > 
> > 
> > it's not ready for accept PR yet
> 
> I will only add missing stuff. e.g the cpp template has
> 
> * deprecrated AXLOG
> * Color4B
> 
> or you can change it of course

thanks, I will do

aismann commented:
its seems alpha is not working well.
Works:
`ax::Color(255, 0, 0, 255)` // alpha = 255 => red line is shown
`ax::Color(255, 0, 0, 0)`     // alpha = 0     => red line is NOT shown

`alpha =100` works not -> seems to be 255:
`ax::Color(255, 0, 0, 100)` // Red should be a little bit transparent too  => but its looks like `ax::Color(255, 0, 0, 255)`

halx99 commented:
the ax::Color same with Color4F must be normalized

aismann commented:
> the ax::Color same with Color4F must be normalized

`normalized `What das this means?
 ax::Color4F(1.0, 0, 0, 0.5) is also no alpha working

halx99 commented:
The ax::Color is just renamed from ax::Color4F, I don't understand what do mean no alpha working

aismann commented:
@halx99 
I checked it again. It works correct. Sorry.
I used ax:Color like ax::Color4B not like ax::Color4F

An `assert >1.0f` is useful?

paulocoutinhox commented on 3rdparty/box2d/CMakeLists.txt:
```diff
@@ -4,105 +4,23 @@ set(target_name ${lib_name})
-	src/dynamics/b2_world.cpp
-	src/dynamics/b2_world_callbacks.cpp
-	src/rope/b2_rope.cpp)
+FILE(GLOB_RECURSE box2d_sources *.h;*.c)
```
put FILE lower case to be compatible with other functions call in cmake

halx99 committed:
All typo: ax::Color4F ==> ax::Color

aismann commented:
I installed axmol on a new machine.
Compiling the 3.0 branch..
Starting the Box2D Testbed and ..boom....
![image](https://github.com/user-attachments/assets/8d7dc409-abd8-4aba-8d11-2fd78e815d5f)

Folder axmol\build\bin\cpp-tests\Debug\axslc\custom has no files:
```
 custom/circle_vs
 custom/circle_fs
```

rh101 commented:
> I installed axmol on a new machine.
> Compiling the 3.0 branch..
> Starting the Box2D Testbed and ..boom....

Given that this is a work in progress, it may be a little premature to be attempting to make use of it for any reason.  Perhaps wait till it is out of draft and ready for testing.

halx99 commented:
> I installed axmol on a new machine. Compiling the 3.0 branch.. Starting the Box2D Testbed and ..boom.... ![image](https://private-user-images.githubusercontent.com/8652787/375983100-8d7dc409-abd8-4aba-8d11-2fd78e815d5f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg3ODk4MTMsIm5iZiI6MTcyODc4OTUxMywicGF0aCI6Ii84NjUyNzg3LzM3NTk4MzEwMC04ZDdkYzQwOS1hYmQ4LTRhYmEtOGQxMS0yZmQ3OGU4MTVkNWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTNUMDMxODMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2Y1NGU4ODJhYTFmZTZiNTg2YjhiZjRmOTc5Y2VjN2I3YmEyMGFiMTU4MTg3ZTFjNmRhOTJkMGI5YzEyNzFmNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.ahOcbFNjGB2YXnLjRUn6nOqCbdjfqWJGM0AQhjR31fI)
> 
> Folder axmol\build\bin\cpp-tests\Debug\axslc\custom has no files:
> 
> ```
>  custom/circle_vs
>  custom/circle_fs
> ```

I foget commit sample shaders to sample-assets repo

halx99 commented:
I already add .fs .vs as glslcc shader file extension: https://github.com/axmolengine/axmol/blob/3.x/cmake/Modules/AXGLSLCC.cmake#L12 

so I don't know why you meet the issue, maybe delete build folder, then re-cmake will fix your issue

aismann commented:
> so I don't know why you meet the issue, maybe delete build folder, then re-cmake will fix your issue

Works now: I download and rebuild axmol again.

aismann commented:
> 
> [...] Perhaps wait till it is out of draft and ready for testing.

@rh101 
You are right for Box2D 3.x stuff.
But this is all done and working for nativ using of Box2D 3.x
=> See the working Box2D tests  on cpp-tests.
All other changes be more like PRs which we doing allways.
And of course this are also working well.

The only alpha status will be the internal physics, but this is not using be me at the moment.

Thats my opinium.

_Edit:_
Last but not least:
Testing this PR in an early state (parts which should work of course, not the alpha part!) permit/detect side effects earlier and can reduce time for the developer, because he needs not so much testing this parts. Of course mistakes done by the 'tester' can be raising too (thats the bad side).

halx99 committed:
Fix LayerGradient not adapte float4 color format

halx99 committed:
Delete unused files

aismann commented:
@halx99 
Please rename also all:
`V2F_C4F_T2F_Triangle` to `V2F_T2F_C4F_Triangle`
`V3F_C4F_T2F_Quad` to `V3F_T2F_C4F_Quad`
`V2F_C4F_T2F_Quad` to `V2F_T2F_C4F_Quad`

smilediver commented:
Why was vertex color changed from bytes to floats? This increases vertex size from 24 bytes to 36 by a 50%! Which will increase memory usage and traffic for no gain and at a performance cost, especially on mobile devices.

halx99 committed:
Fix preprocessor checks

halx99 committed:
Update openal-soft to 1.24.1

halx99 commented:
> Why was vertex color changed from bytes to floats? This increases vertex size from 24 bytes to 36 by a 50%! Which will increase memory usage and traffic for no gain and at a performance cost, especially on mobile devices.

Will evaluating long-term.

halx99 commented:
> @halx99 Please rename also all: `V2F_C4F_T2F_Triangle` to `V2F_T2F_C4F_Triangle` `V3F_C4F_T2F_Quad` to `V3F_T2F_C4F_Quad` `V2F_C4F_T2F_Quad` to `V2F_T2F_C4F_Quad`

done

halx99 committed:
Fix #2254

halx99 committed:
Rename all XXX_C4F_T2F_XXX to XXX_T2F_C4F_XXX

halx99 committed:
Still use V3F_T2F_C4B for sprite render

halx99 committed:
Update lua bindings

halx99 committed:
Merge axmol dev into 3.x

rh101 committed:
AxmolActivity will extend AppCompatActivity
Min SDK set to 21
Tests updated to support AppCompatActivity
MultiDexEnabled for cpp-tests

halx99 committed:
Update Android & iOS minmal OS support

Android: Android 5.1(API 21)+
iOS: 13+

halx99 committed:
Update simdjson to v3.11.4

halx99 committed:
Remove dependency rapidjson for JsonWriter

paulocoutinhox commented:
RapidJson can be removed from version 3 and Jeep only nloman lib. What do you think?

halx99 commented:
API : nloman ~= boost.json ~= simdjson >> rapidjson
speed: simdjson(readonly) > yyjson >> rapidjson ~= boost.json >> nloman

halx99 commented:
For speed: the simdjson still is best choice
For dom operation: boost.json would be better than nolman, because it's API easy to use and speed better than nolman
For json emitter, axmol have `JsonWriter`, and speed would equal to yyjson in release build

johnspeny commented:
I use rapid json, what is it better at than the other jsons listed?  So are you really removing it

rh101 commented:
> I use rapid json, what is it better at than the other jsons listed? So are you really removing it

If you still need it after (or if) it's removed from the engine, then just add it to your own project.  The same goes for any other third-party library.

halx99 committed:
Update openal-soft to 1.24.2

halx99 committed:
Update simdjson to v3.11.5

halx99 committed:
Fix #2330 and improve code style

halx99 committed:
Improve code style

halx99 committed:
Fix linux link error

halx99 committed:
Rename Color4B to Color32

halx99 committed:
Fix wasm64 build

halx99 committed:
Fix wasm64 runtime exception for imgui glfw backend

halx99 committed:
Add wasm64 build to ci

halx99 committed:
Improve Console::Utility::trim and move to StringUtils

    - Use c++20 reature ranges
    - Use c++20 0-GC split to instead std::getline

halx99 committed:
Add lib: fast_float

-  Resolve clang libc++ not support std::from_chars for floating points

halx99 committed:
Improve c3t, c3b load

- Use simdjson to load c3t
- Use yasio::ibstream to c3b

aismann commented:
> Rename Color4B to Color32

@halx99 
Is it not better to have Color4B (self explaining)
And what about Color3B and others?

halx99 committed:
Fix android appcompat not set in template project

halx99 committed:
Fix wasm compiling warnigs and errors

halx99 committed:
Update simdjson and fast_float
